service/src/main/java/org/cbioportal/service/impl/GenePanelServiceImpl.java

```diff
@@ -172,37 +178,106 @@ public GenePanel getGenePanel(String genePanelId) throws GenePanelNotFoundExcept
                     resultGenePanelData.setProfiled(true);
                 } else {
-                    resultGenePanelData.setProfiled(false);
+                    List<SampleList> sampleLists = sampleListMap.get(studyId);
+                    if (molecularProfile.getMolecularAlterationType().equals(MolecularProfile.MolecularAlterationType.MUTATION_EXTENDED) && 
+                        (sampleLists == null || (sampleLists != null && sampleLists.get(0).getSampleIds().contains(sampleId)))) {
+                        resultGenePanelData.setProfiled(true);
+                    } else {
+                        resultGenePanelData.setProfiled(false);
+                    }
                 }
-            }
-            resultGenePanelData.setMolecularProfileId(molecularProfileId);
-            resultGenePanelData.setSampleId(sampleId);
-            Optional<Sample> sample = Optional.ofNullable((Sample)samples.get(sampleId, studyId));
-            if (sample.isPresent()) {
-                resultGenePanelData.setPatientId(sample.get().getPatientStableId());
+                resultGenePanelData.setMolecularProfileId(molecularProfileId);
+                resultGenePanelData.setSampleId(sampleId);
+                
+                Sample sample = (Sample)samples.get(sampleId, studyId);
+                resultGenePanelData.setPatientId(sample.getPatientStableId());
                 resultGenePanelDataList.add(resultGenePanelData);
             }
         }
 
         return resultGenePanelDataList;
     }
 
+    private List<GenePanelData> createGenePanelDatabyPatientIds(
+            MultiKeyMap genePanelDataMap,
+            List<String> molecularProfileIds,
+            List<String> patinetIds) {
+
+        Map<String, MolecularProfile> molecularProfileMap = molecularProfileService
+                .getMolecularProfiles(new ArrayList<>(new HashSet<>(molecularProfileIds)), "SUMMARY")
+                .stream()
+                .collect(Collectors.toMap(MolecularProfile::getStableId, Function.identity()));
+        
+        List<String> studyIds = molecularProfileIds
+                .stream()
+                .map(molecularProfileMap::get) // assuming all molecular profiles are valid. invalid id would throw runtime exception
+                .map(MolecularProfile::getCancerStudyIdentifier)
+                .collect(Collectors.toList());
+
+        Map<Pair<String, String>, List<Sample>> sampleSet = createSampleMapByPatientId(studyIds, patinetIds);
+
+        List<String> sequencedSampleListIds = studyIds
+                .stream()
+                .distinct()
+                .map(studyId -> studyId + SEQUENCED_LIST_SUFFIX)
+                .collect(Collectors.toList());
+
+        Map<String, List<SampleList>> sampleListMap = sampleListService
+                .fetchSampleLists(sequencedSampleListIds, "DETAILED")
+                .stream()
+                .collect(Collectors.groupingBy(SampleList::getCancerStudyIdentifier));
```

Comment:
Looks like a lot of copy paste code from above, intermingled with some new objects. Would be nice to pull out the duplicate stuff into a single function.

Reply:
Yeah, I noticed it too. I'll do it in another pr